File: peqrouter/scenarios.py
"""Telemetry parsing and built-in demonstration scenarios."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any, Mapping

from .models import (
    CLOUD_AI_100_TDP_WATTS,
    Device,
    DeviceConfig,
    DeviceTelemetry,
    ValidationError,
    default_device_configs,
)

# Measured medians from the 106-prompt calibration sweep
# (benchmarks/calibration/heads/heads.json -> device_constants). Energy: 106
# samples each for phone/PC. Decode throughput: same source, also 106 samples
# each -- phone's model is much smaller (Qwen3-0.6B) and much faster than
# PC's (Qwen3-VL-4B-Instruct on the X-Elite NPU), so PC is both slower AND
# more energy-hungry per token, not a wash.
_PHONE_JOULES_PER_TOKEN = 0.0412
_PHONE_TOKENS_PER_SECOND = 94.22
_PC_JOULES_PER_TOKEN = 0.4447
_PC_TOKENS_PER_SECOND = 20.19

# Cloud has no decode-only throughput or per-token energy constant the way
# phone/PC do (heads.json reports both null for cloud -- CirrascaleExecutor
# never captured ttft/decode-speed, only whole-call latency and token
# counts). What *is* measured, from benchmarks/calibration/runs/
# sweep_cloud_llama70b.jsonl (100 rows with both fields): the median
# end-to-end rate of (prompt_tokens + completion_tokens) / api_latency_ms
# across the sweep, folding network + queueing + generation into one number
# since they were never captured separately for cloud. Because this rate
# already includes network time, cloud's network_latency_ms below is 0 --
# adding a separate network hop on top would double-count it. This is the
# same "wall-clock, not compute-only" tradeoff CLOUD_AI_100_TDP_WATTS makes
# for cloud's energy, for the same reason.
_CLOUD_TOKENS_PER_SECOND = 33.37
_CLOUD_JOULES_PER_TOKEN = CLOUD_AI_100_TDP_WATTS / _CLOUD_TOKENS_PER_SECOND


def _healthy() -> dict[Device, DeviceTelemetry]:
    return {
        # network_latency_ms for phone/PC (8/18) is still an illustrative,
        # unmeasured LAN-hop placeholder -- nothing in this repo measures
        # cross-device network latency. Left as-is rather than invented.
        Device.PHONE: DeviceTelemetry(
            True, 8, _PHONE_TOKENS_PER_SECOND, _PHONE_JOULES_PER_TOKEN, 0.08, 0.08, 95, 0
        ),
        Device.PC: DeviceTelemetry(
            True, 18, _PC_TOKENS_PER_SECOND, _PC_JOULES_PER_TOKEN, 0.18, 0.15, 85, 0
        ),
        Device.CLOUD: DeviceTelemetry(
            True, 0, _CLOUD_TOKENS_PER_SECOND, _CLOUD_JOULES_PER_TOKEN, 0.20, 0.12, None, 0.03
        ),
    }


def built_in_scenarios() -> dict[str, dict[Device, DeviceTelemetry]]:
    healthy = _healthy()
    phone_low_battery = dict(healthy)
    phone_low_battery[Device.PHONE] = DeviceTelemetry(
        True, 8, _PHONE_TOKENS_PER_SECOND, _PHONE_JOULES_PER_TOKEN, 0.08, 0.08, 4, 0
    )
    pc_congested = dict(healthy)
    pc_congested[Device.PC] = DeviceTelemetry(
        # Congestion is modeled as half the healthy throughput and ~2x the
        # healthy energy draw -- same ratios as before the healthy baseline
        # was corrected to the measured values.
        True, 18, _PC_TOKENS_PER_SECOND / 2, _PC_JOULES_PER_TOKEN * 2, 0.92, 0.90, 75, 0
    )
    cloud_offline = dict(healthy)
    cloud_offline[Device.CLOUD] = DeviceTelemetry(
        False, 0, 0, _CLOUD_JOULES_PER_TOKEN, 0.0, 0.0, None, 0.03
    )
    return {
        "healthy": healthy,
        "phone-low-battery": phone_low_battery,
        "pc-congested": pc_congested,
        "cloud-offline": cloud_offline,
    }


def telemetry_from_mapping(data: Mapping[str, Any]) -> dict[Device, DeviceTelemetry]:
    expected = {device.value for device in Device}
    if set(data) != expected:
        raise ValidationError("telemetry JSON must contain exactly phone, pc, and cloud")
    result: dict[Device, DeviceTelemetry] = {}
    for device in Device:
        values = data[device.value]
        if not isinstance(values, Mapping):
            raise ValidationError(f"telemetry for {device.value} must be a JSON object")
        try:
            result[device] = DeviceTelemetry(**values)
        except TypeError as error:
            raise ValidationError(f"invalid telemetry fields for {device.value}: {error}") from error
    return result


def load_telemetry(path: str | Path) -> dict[Device, DeviceTelemetry]:
    with Path(path).open(encoding="utf-8") as stream:
        data = json.load(stream)
    if not isinstance(data, Mapping):
        raise ValidationError("telemetry JSON root must be an object")
    return telemetry_from_mapping(data)


def device_configs_from_mapping(data: Mapping[str, Any]) -> dict[Device, DeviceConfig]:
    device_data = data.get("devices")
    if not isinstance(device_data, Mapping):
        raise ValidationError("model configuration must contain a devices object")
    expected = {device.value for device in Device}
    if set(device_data) != expected:
        raise ValidationError("model configuration must contain exactly phone, pc, and cloud")
    configs = default_device_configs()
    for device in Device:
        values = device_data[device.value]
        if not isinstance(values, Mapping):
            raise ValidationError(f"configuration for {device.value} must be a JSON object")
        try:
            configs[device] = DeviceConfig(**values)
        except TypeError as error:
            raise ValidationError(f"invalid configuration fields for {device.value}: {error}") from error
    return configs


def load_device_configs(path: str | Path) -> dict[Device, DeviceConfig]:
    with Path(path).open(encoding="utf-8") as stream:
        data = json.load(stream)
    if not isinstance(data, Mapping):
        raise ValidationError("model configuration JSON root must be an object")
    return device_configs_from_mapping(data)

File: s25_android_app/src/main/java/com/geniex/demo/router/Scenarios.kt
// ---------------------------------------------------------------------
// Copyright (c) 2026 Qualcomm Technologies, Inc. and/or its subsidiaries.
// SPDX-License-Identifier: BSD-3-Clause
// ---------------------------------------------------------------------
package com.geniex.demo.router

/**
 * Kotlin port of `ecorouter/scenarios.py`'s built-in telemetry scenarios.
 *
 * Live telemetry collection (battery/thermal/network probes on-device) is
 * not implemented here yet, same as it isn't on the PC side -- see
 * TODO.md's "Connect live telemetry collectors" item. Until then, routing
 * from the phone picks one of these canned snapshots the same way the PC
 * CLI's `--scenario` flag does.
 */
object Scenarios {

    // Measured medians from the 106-prompt calibration sweep
    // (benchmarks/calibration/heads/heads.json -> device_constants) -- kept
    // numerically identical to scenarios.py so on-device and PC routing agree
    // on a given scenario.
    private const val PHONE_JOULES_PER_TOKEN = 0.0412
    private const val PHONE_TOKENS_PER_SECOND = 94.22
    private const val PC_JOULES_PER_TOKEN = 0.4447
    private const val PC_TOKENS_PER_SECOND = 20.19
    private const val CLOUD_TOKENS_PER_SECOND = 33.37
    private val CLOUD_JOULES_PER_TOKEN = CLOUD_AI_100_TDP_WATTS / CLOUD_TOKENS_PER_SECOND

    private fun healthy(): Map<Device, DeviceTelemetry> = mapOf(
        Device.PHONE to DeviceTelemetry(
            available = true,
            networkLatencyMs = 8.0,
            throughputTokensPerSecond = PHONE_TOKENS_PER_SECOND,
            energyJoulesPerToken = PHONE_JOULES_PER_TOKEN,
            utilization = 0.08,
            thermalPressure = 0.08,
            batteryPercent = 95.0,
            cloudCostPer1kTokensUsd = 0.0,
        ),
        Device.PC to DeviceTelemetry(
            available = true,
            networkLatencyMs = 18.0,
            throughputTokensPerSecond = PC_TOKENS_PER_SECOND,
            energyJoulesPerToken = PC_JOULES_PER_TOKEN,
            utilization = 0.18,
            thermalPressure = 0.15,
            batteryPercent = 85.0,
            cloudCostPer1kTokensUsd = 0.0,
        ),
        Device.CLOUD to DeviceTelemetry(
            available = true,
            networkLatencyMs = 0.0,
            throughputTokensPerSecond = CLOUD_TOKENS_PER_SECOND,
            energyJoulesPerToken = CLOUD_JOULES_PER_TOKEN,
            utilization = 0.20,
            thermalPressure = 0.12,
            batteryPercent = null,
            cloudCostPer1kTokensUsd = 0.03,
        ),
    )

    fun builtInScenarios(): Map<String, Map<Device, DeviceTelemetry>> {
        val healthy = healthy()

        val phoneLowBattery = healthy.toMutableMap()
        phoneLowBattery[Device.PHONE] = healthy.getValue(Device.PHONE).copy(batteryPercent = 4.0)

        val pcCongested = healthy.toMutableMap()
        // Congestion modeled as half the healthy throughput and ~2x the
        // healthy energy draw, same ratios as scenarios.py.
        pcCongested[Device.PC] = healthy.getValue(Device.PC).copy(
            throughputTokensPerSecond = PC_TOKENS_PER_SECOND / 2,
            energyJoulesPerToken = PC_JOULES_PER_TOKEN * 2,
            utilization = 0.92,
            thermalPressure = 0.90,
            batteryPercent = 75.0,
        )

        val cloudOffline = healthy.toMutableMap()
        cloudOffline[Device.CLOUD] = healthy.getValue(Device.CLOUD).copy(
            available = false,
            networkLatencyMs = 0.0,
            throughputTokensPerSecond = 0.0,
            utilization = 0.0,
            thermalPressure = 0.0,
        )

        return mapOf(
            "healthy" to healthy,
            "phone-low-battery" to phoneLowBattery,
            "pc-congested" to pcCongested,
            "cloud-offline" to cloudOffline,
        )
    }
}
